### PR TITLE
Fix issue 22770: C++ header generator generates trailing newlines

### PR DESCRIPTION
**Pull request by ljmf00**
This patch adds logic to check for trailing newlines in dtoh generator and
remove them.

Signed-off-by: Luís Ferreira <contact@lsferreira.net>

---

Related to #13638 .

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/dtoh.d [hunk 1]
```diff
@@ -180,6 +180,21 @@ struct _d_dynamicArray final
 `);
     }
 
+    // prevent trailing newlines
+    foreach_reverse (i, ch; buf)
```
Comment (ljmf00) [inline src/dmd/dtoh.d:184]: Maybe this is overkill logic. Checking for `['\n', '\n']` should be enough, but if any node introduces double newlines, this is required.

**Comment by ljmf00** [comment]
@MoonlightSentinel you self-requested review but didn't leave a comment, perhaps you left it pending?

File: src/dmd/dtoh.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dmd/dtoh.d:184]: Check for repeated newlines (either `\r\n`  or `\n`) should be sufficient

File: src/dmd/dtoh.d [hunk 2]
```diff
@@ -180,6 +180,21 @@ struct _d_dynamicArray final
...
+    {
+        if (ch != '\n')
```
Comment (MoonlightSentinel) [inline src/dmd/dtoh.d:186]: Produces invalid line ending on windows

File: src/dmd/dtoh.d [hunk 3]
```diff
@@ -180,6 +180,21 @@ struct _d_dynamicArray final
...
+        {
+            if (buf.length)
+            {
+                size_t count = (buf.length - 1) - i;
+                if (count)
+                    buf.remove(buf.length - 1, count - 1);
```
Comment (MoonlightSentinel) [inline src/dmd/dtoh.d:192]: Iteration + modification will work with the current `Outbuffer` implementation but it's probably cleaner to determine the proper end index + resize at the end.

**Review by MoonlightSentinel — CHANGES_REQUESTED** [review]

File: src/dmd/dtoh.d (hunk 3 above)
Comment (ljmf00) [reply src/dmd/dtoh.d:192]: This will not iterate, when `if (ch != '\n')` is triggered, the code will only run once.

File: src/dmd/dtoh.d (hunk 1 above)
Comment (ljmf00) [reply src/dmd/dtoh.d:184]: You mean, specifically only twice or more than one? I don't see any case where dtoh generates `\n\n\n`, so maybe the twice case is sufficient.

Comment (MoonlightSentinel) [reply src/dmd/dtoh.d:184]: More than one. I don't expect more than two but just to be sure.

**Comment by ljmf00** [comment]
Seems like my Sourcegraph plugin is telling me that Windows builds are not generating coverage 
image

**Comment by ljmf00** [comment]
> ping @ljmf00

Thanks for pinging. @MoonlightSentinel can you re-review?

File: src/dmd/common/outbuffer.d [hunk 4]
```diff
@@ -662,6 +662,11 @@ struct OutBuffer
         return cast(char)data[i];
     }
 
+    extern (D) size_t opDollar() const pure nothrow @nogc @safe
+    {
+        return length();
+    }
```
Comment (MoonlightSentinel) [inline src/dmd/common/outbuffer.d:668]:
```suggestion
    alias opDollar = length;
```

**Review by MoonlightSentinel — APPROVED** [review]
Otherwise LGTM

**Comment by MoonlightSentinel** [comment]
> Seems like my Sourcegraph plugin is telling me that Windows builds are not generating coverage

Weird. The coverage files includes those lines on windows and the corresponding file [was uploaded to CodeCov](https://dev.azure.com/dlanguage/dmd/_build/results?buildId=28627&view=logs&j=2e8e4583-3493-5411-a5ed-6353b3fd9e1d&t=f29a0a4f-1685-5b45-b0f4-a796fba54562&l=223)

**Comment by ljmf00** [comment]
> > Seems like my Sourcegraph plugin is telling me that Windows builds are not generating coverage
> 
> Weird. The coverage files includes those lines on windows and the corresponding file [was uploaded to CodeCov](https://dev.azure.com/dlanguage/dmd/_build/results?buildId=28627&view=logs&j=2e8e4583-3493-5411-a5ed-6353b3fd9e1d&t=f29a0a4f-1685-5b45-b0f4-a796fba54562&l=223)

According to codecov https://app.codecov.io/gh/dlang/dmd/branch/dtoh-fix-trailing-new-line the coverage decreased a lot on my change. Probably worth investigating it.

**Pull request merged**